In [ ]:
import pandas as pd
df = pd.read_excel('Online retail.xlsx')
df

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                                burgers,meatballs,eggs                                                                                                                                                                             
1                                               chutney                                                                                                                                                                             
2                                        turkey,avocado                                                                                                                                                                             
3     mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                        low fat yogurt                                                                                                                                                                             
...                                                 ...                                                                                                                                                                             
7495                      butter,light mayo,fresh bread                                                                                                                                                                             
7496  burgers,frozen vegetables,eggs,french fries,ma...                                                                                                                                                                             
7497                                            chicken                                                                                                                                                                             
7498                                 escalope,green tea                                                                                                                                                                             
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt                                                                                                                                                                             

[7500 rows x 1 columns]

In [ ]:
#renaming the coloumn to 'Transaction' for clarity
df.columns= ['Transaction']

In [ ]:
df.isnull().sum()

Transaction    0
dtype: int64

In [ ]:
# Convert the transactions to a list of lists
transaction = df['Transaction'].apply(lambda x: x.split(','))

In [ ]:
#Convert the list of transaction into a one-hot encoded Dataframe
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_ary = te.fit(transaction).transform(transaction)
df = pd.DataFrame(te_ary,columns = te.columns_)

df.head()

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False    False              False      False    False        False   
1       False    False              False      False    False        False   
2       False    False              False      False     True        False   
3       False    False              False      False    False        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False           False   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...    True           False   
3  False           False      False        False  ...   False           False   
4  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False             False              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False  False        False     False  
1             False  False        False     False  
2             False  False        False     False  
3              True  False        False     False  
4             False  False        False     False  

[5 rows x 120 columns]

In [ ]:
from mlxtend.frequent_patterns import apriori,association_rules

In [ ]:
#apply the Apriori algorithm to find frequent itemsets
frequent_itemsets = apriori(df, min_support=0.01, use_colnames=True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Generate the association rules from the frequent itemsets
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1.0)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Sort rules by confidence in descending order
rules = rules.sort_values(by='confidence', ascending=False)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
#Display the top 10 rules
rules.head(10)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents      consequents  antecedent support  \
344        (eggs, ground beef)  (mineral water)            0.020000   
378        (milk, ground beef)  (mineral water)            0.022000   
320   (chocolate, ground beef)  (mineral water)            0.023067   
366  (milk, frozen vegetables)  (mineral water)            0.023600   
274                     (soup)  (mineral water)            0.050533   
402      (pancakes, spaghetti)  (mineral water)            0.025200   
396     (olive oil, spaghetti)  (mineral water)            0.022933   
390          (milk, spaghetti)  (mineral water)            0.035467   
326          (milk, chocolate)  (mineral water)            0.032133   
384   (spaghetti, ground beef)  (mineral water)            0.039200   

     consequent support   support  confidence      lift  leverage  conviction  \
344            0.238267  0.010133    0.506667  2.126469  0.005368    1.544054   
378            0.238267  0.011067    0.503030  2.111207  0.005825    1.532756   
320            0.238267  0.010933    0.473988  1.989319  0.005437    1.448130   
366            0.238267  0.011067    0.468927  1.968075  0.005444    1.434328   
274            0.238267  0.023067    0.456464  1.915771  0.011026    1.401441   
402            0.238267  0.011467    0.455026  1.909736  0.005462    1.397744   
396            0.238267  0.010267    0.447674  1.878880  0.004802    1.379138   
390            0.238267  0.015733    0.443609  1.861817  0.007283    1.369061   
326            0.238267  0.014000    0.435685  1.828559  0.006344    1.349836   
384            0.238267  0.017067    0.435374  1.827256  0.007727    1.349094   

     zhangs_metric  
344       0.540548  
378       0.538177  
320       0.509058  
366       0.503778  
274       0.503458  
402       0.488682  
396       0.478747  
390       0.479911  
326       0.468165  
384       0.471202

Common Consequent Item:

All the rules have "mineral water" as the consequent item, indicating that mineral water is frequently associated with various other products.

High Confidence Rules:

The rules show confidence values between approximately 0.44 to 0.51, indicating that when the antecedent items are purchased, there is about a 44% to 51% chance that mineral water will also be purchased.

Lift Values:

The lift values range from approximately 1.86 to 2.13. A lift value greater than 1 indicates a positive correlation between the antecedent and consequent. Here, it suggests that the purchase of the antecedent items increases the likelihood of purchasing mineral water by about 1.86 to 2.13 times compared to random chance.

Many rules indicate that customers purchasing protein-rich foods (e.g., eggs, ground beef), essential groceries (e.g., milk, frozen vegetables), or items associated with healthy cooking (e.g., olive oil) are also buying mineral water. This suggests that these customers are health-conscious and prefer to stay hydrated with mineral water.

**Interview Questions:**

1.	What is lift and why is it important in Association rules?

Lift Definition:

Lift is a measure used in association rule mining to evaluate the strength of a rule over the baseline probability of the consequent item. It is defined as the ratio of the observed support of the rule to the expected support if the antecedent and consequent were independent. Mathematically, it is expressed as:

Lift
(
𝐴
→
𝐵
)
=
Support
(
𝐴
∪
𝐵
)
Support
(
𝐴
)
×
Support
(
𝐵
)
Lift(A→B)=
Support(A)×Support(B)
Support(A∪B)
​

Why Lift is Important in Association Rules:

Evaluates Rule Strength:

Lift quantifies how much more likely the consequent item is to be purchased when the antecedent item(s) are purchased, compared to random chance. A lift value greater than 1 indicates a positive association, meaning that the presence of the antecedent increases the likelihood of the consequent.

Identifies Interesting Rules:

By examining lift, we can identify the most interesting and potentially valuable rules. Higher lift values suggest stronger associations between items, which can be crucial for decision-making in marketing, sales, and inventory management.

Reduces Noise from Common Items:

Some items might have high support due to their popularity, but they may not provide significant insights if they are common in many transactions. Lift helps filter out such noise by normalizing the support with the baseline probability.

Helps in Cross-Selling and Up-Selling:

Rules with high lift values can be used to design effective cross-selling and up-selling strategies. For example, if the lift value for "milk" and "cookies" is high, retailers can place these items together or create bundle offers.

Informs Product Placement:

Lift can guide product placement in stores. Items with strong associations (high lift values) can be placed near each other to increase the likelihood of joint purchases, enhancing the customer shopping experience and increasing sales.

Improves Marketing Campaigns:

Marketing campaigns can be tailored based on lift values. Products that have high lift values with certain antecedents can be featured together in advertisements, improving the campaign's effectiveness by targeting likely purchasing behaviors.


2.	What is support and Confidence. How do you calculate them?


Support and Confidence are two fundamental metrics used in association rule mining to measure the frequency and reliability of rules.

1. Support:

Definition:
Support measures how frequently the items in an itemset appear together in the dataset. It is the proportion of transactions in the dataset that contain the itemset.


2. Confidence:

Definition:
Confidence measures how often items in the consequent (e.g., B) appear in transactions that contain the antecedent (e.g., A). It is the conditional probability of the consequent given the antecedent.

Summary:
Support quantifies the popularity of an itemset by indicating how frequently it occurs in the dataset.
Confidence evaluates the reliability of the rule by showing the likelihood of the consequent given that the antecedent has occurred.
These metrics help in identifying interesting rules and patterns within the dataset, enabling better decision-making for marketing, product placement, and inventory management.









3.	What are some limitations or challenges of Association rules mining?

Association rule mining is a powerful technique for discovering relationships between items in large datasets, but it comes with several limitations and challenges:

1. Scalability and Performance:

Challenge:

As the dataset size grows, the number of possible itemsets and rules increases exponentially. This can lead to high computational costs and memory usage.

Example:

In a supermarket dataset with thousands of products, generating all possible itemsets and rules can become infeasible.

2. Meaningful Rule Extraction:

Challenge:

Not all generated rules are meaningful or useful. A large number of rules might be generated, many of which may have low support and confidence, making it difficult to identify the truly valuable ones.

Example:

Rules like
{
𝑏
𝑟
𝑒
𝑎
𝑑
}
→
{
𝑏
𝑢
𝑡
𝑡
𝑒
𝑟
}
{bread}→{butter} might be overshadowed by many insignificant rules.

3. Setting Appropriate Thresholds:

Challenge:

Choosing the right support and confidence thresholds is critical but challenging. Too high thresholds might miss important rules, while too low thresholds can result in an overwhelming number of rules.

Example:

Setting a high minimum support might eliminate interesting rare item associations, such as niche product combinations that a small but significant segment of customers buys.

4. Interpretability:

Challenge:

Some rules can be complex and hard to interpret, especially when involving multiple items.
Example:
A rule like
{
𝑚
𝑖
𝑙
𝑘
,
𝑏
𝑟
𝑒
𝑎
𝑑
,
𝑏
𝑢
𝑡
𝑡
𝑒
𝑟
}
→
{
𝑗
𝑎
𝑚
}
{milk,bread,butter}→{jam} might be less intuitive to understand and act upon compared to simpler rules.

5. Redundancy and Overfitting:

Challenge:

Many rules can be redundant or represent overfitting, where the rules are too specific to the dataset and may not generalize well to new data.

Example:

Rules derived from seasonal data might not apply to other




